rq/batch.py

```diff
@@ -0,0 +1,85 @@
+import logging
+from typing import List, Union, Optional
+from uuid import uuid4
+
+from redis import Redis
+from redis.client import Pipeline
+
+from .job import Job
+from .utils import as_text
+
+logger = logging.getLogger("rq.job")
+
+
+class Batch:
+    REDIS_BATCH_NAME_PREFIX = 'rq:batch:'
+
+    def __init__(
+        self, id: str = str(uuid4())[-12:], jobs: List[Job] = None, connection: Optional['Redis'] = None, ttl=500
+    ):
+        self.id = id
+        self.connection = connection
+        self.key = '{0}{1}'.format(self.REDIS_BATCH_NAME_PREFIX, self.id)
+        self.jobs_key = self.key + ":jobs"
+        self.ttl = ttl
+
+        self.jobs = []
+        if jobs:
+            self.add_jobs(jobs)
+
+    def add_jobs(self, jobs: List[Union['Job', str]], pipeline=None):
+        pipe = pipeline if pipeline else self.connection.pipeline
+        self.job_ids = [job.id for job in jobs]
+        self.connection.sadd(self.jobs_key, *self.job_ids)
+        self.jobs += jobs
+        for job in jobs:
+            job.set_batch_id(self.id)
+            job.save(pipeline=self.connection)
+        self.renew_ttl()
+        self.save()
+
+    def fetch_jobs(self) -> list:
+        self.job_ids = [as_text(job) for job in self.connection.smembers(self.key + ":jobs")]
+        self.jobs = Job.fetch_many(self.job_ids, self.connection)
+
+    def renew_ttl(self, pipeline=None):
```

Comment:
I don't think we should worry about Batch TTLs, they should always be persisted. We can simply delete the batch key once the last job in batch is deleted. What do you think?

Reply:
Reason I went with the Batch TTL is I thought it might be expensive to determine whether a job is the "last" job in a large batch. If the job is expired in Redis, I assume we would need workers to run a maintenance task to check if any of the batch's jobs still exist. Lmk if I'm off here.

Also, for our use case, we'd ideally be able to see the status of all jobs in a batch as an atomic unit. If one user enqueues 100 jobs in a batch, and some expire before the whole batch has finished, it could mislead another user checking the status of the jobs. That's why the batch TTL that is inherited by the member jobs was appealing,

Reply:
RQ runs maintenance tasks every 10 minutes or so. https://github.com/rq/rq/blob/10230ff0404073d892418c04879d8f04ad57ca9e/rq/worker.py#L309 . This is where we remove expired/deleted jobs from registries. So the mechanism for this already exists.

> it might be expensive to determine whether a job is the "last" job in a large batch.
I think we can around this by batching the deletions. For example if we have 1000 expired jobs in a batch of a million jobs, we can remove 1,000 jobs in a single call and simply check how many jobs are left in the queue before deleting it. 

I just checked Redis' set documentation and it just so happens that the way sets are implemented in Redis, when the last element is deleted from the set, the key will also be deleted. https://redis.io/commands/srem/

Reply:
What are your thoughts on having the batch TTL be optional and determine the job cleanup method? If batch_ttl is None, the jobs behave as normal with their own TTLs, and the maintenance task eventually deletes the jobs from the batch:jobs set when the job is deleted from the registry. If the user specifies a TTL, the `persist_jobs` and `renew_ttl` methods are used so all of the batches jobs expire at about the same time after the last job completes, which is useful if the user doesn't want to deal with partial batches.

Reply:
> What are your thoughts on having the batch TTL be optional and determine the job cleanup method?

Actually, I think `Batch` objects themselves should not have TTLs. `Batch` should be just a light way of grouping jobs together. Each job within a batch can have their own TTLs. If needed, we can provide a shortcut to change TTLs of all jobs within a batch.

> If batch_ttl is None, the jobs behave as normal with their own TTLs, and the maintenance task eventually deletes the jobs from the batch:jobs set when the job is deleted from the registry.

Yes, I think this is the way to go.

Additionally, we'll need a key to store all available batches, similar to how all workers are stored in a single key https://github.com/rq/rq/blob/master/rq/worker.py#L245

Reply:
> This is definitely something we'd like, is the current method of updating all job TTLs in a batch when a job completes acceptable?

Each job's TTL should be managed by the user, no implicit TTL updates should be done. What we need to manage are just the batch object themselves.